File: companies/Pinterest/77.combinations.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        /*
        Approach:
        Backtrack increasing numbers so every combination is built once. Prune
        the loop when there are not enough remaining numbers to fill k slots.

        Complexity: O(C(n,k) * k) time and O(k) recursion space excluding output.
        */
        vector<vector<int>> ans;
        vector<int> path;
        function<void(int)> dfs = [&](int start) {
            if ((int)path.size() == k) {
                ans.push_back(path);
                return;
            }
            int need = k - (int)path.size();
            for (int value = start; value <= n - need + 1; ++value) {
                path.push_back(value);
                dfs(value + 1);
                path.pop_back();
            }
        };
        dfs(1);
        return ans;
    }
};

File: companies/Bloomberg/77.combinations.java
import java.util.*;

/**
 * Algorithm:
 * Backtrack combinations in increasing order. The start value prevents
 * duplicate orderings, and the loop bound prunes branches with too few numbers
 * remaining to complete k values.
 *
 * Complexity:
 * Time O(C(n,k) * k), recursion space O(k), plus output.
 */
class Solution {
    private int n;
    private int k;
    private List<Integer> path;
    private List<List<Integer>> ans;

    public List<List<Integer>> combine(int n, int k) {
        this.n = n;
        this.k = k;
        path = new ArrayList<>();
        ans = new ArrayList<>();
        dfs(1);
        return ans;
    }

    private void dfs(int start) {
        if (path.size() == k) {
            ans.add(new ArrayList<>(path));
            return;
        }
        int need = k - path.size();
        for (int value = start; value <= n - need + 1; value++) {
            path.add(value);
            dfs(value + 1);
            path.remove(path.size() - 1);
        }
    }
}

